### PR TITLE
Add Virtual Router top level resource. Add Identity under Microsoft.Network/ExpressRoutePort and MacSecConfig under Microsoft.Network/ExpressRouteLink

### PR DESCRIPTION
**Pull request by naveenchekuri**
If you are a MSFT employee you can view your [work branch via this link](https://portal.azure-devex-tools.com/branch/naveenchekuri/azure-rest-api-specs/network-august-release...master/).

### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/expressRoutePort.json [hunk 1]
```diff
@@ -563,6 +563,32 @@
         }
       }
     },
+    "ExpressRouteLinkMacSecConfig": {
```
Comment (lirenhe) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/expressRoutePort.json]: As you are adding new property to the response, is this a breaking change to the existing API?

**Comment by number213** [comment]
Updated commit to the next API version (2019-07-01) and re-targeted to `network-august-release`

File: specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/expressRoutePort.json (hunk 1 above)
Comment (naveenchekuri) [reply specification/network/resource-manager/Microsoft.Network/stable/2019-06-01/expressRoutePort.json]: I'm adding new properties. This is not a breaking changes as the API can be called by only one customer. If the customer is using updated SDK then can set and get the new fields. If not, customer can use the API without setting or Getting new fields

File: specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json [hunk 2]
```diff
@@ -0,0 +1,699 @@
...
+          "Get VirtualRouter": {
+            "$ref": "./examples/VirtualRouterGet.json"
+          }
+        }
+      },
+      "patch": {
+        "tags": [
+          "VirtualRouters"
+        ],
+        "operationId": "VirtualRouters_UpdateTags",
```
Comment (sanjaiganesh) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json]: >VirtualRouters_UpdateTags [](start = 24, length = 25)

Just  VirtualRouters_Update, as we could always expand this to patch other properties too

File: specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json [hunk 3]
```diff
@@ -0,0 +1,699 @@
...
+          },
+          {
+            "name": "$expand",
+            "in": "query",
+            "required": false,
+            "type": "string",
+            "description": "Expands referenced resources."
+          }
+        ],
+        "responses": {
```
Comment (sanjaiganesh) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json]: >responses [](start = 9, length = 9)

default response missing

File: specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json [hunk 4]
```diff
@@ -0,0 +1,699 @@
...
+          },
+          {
+            "$ref": "./network.json#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "./network.json#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
+          "202": {
```
Comment (sanjaiganesh) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json]: >202 [](start = 11, length = 3)

default response missing

File: specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json [hunk 5]
```diff
@@ -0,0 +1,699 @@
...
+            "description": "Parameters supplied to the create or update Virtual Router."
+          },
+          {
+            "$ref": "./network.json#/parameters/ApiVersionParameter"
+          },
+          {
+            "$ref": "./network.json#/parameters/SubscriptionIdParameter"
+          }
+        ],
+        "responses": {
```
Comment (sanjaiganesh) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json]: >responses [](start = 9, length = 9)

default response missing

Comment (sanjaiganesh) [reply specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json]: same in other apis

---
In reply to: [317344690](https://github.com/Azure/azure-rest-api-specs/pull/7005#discussion_r317344690) [](ancestors = 317344690)

File: specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json [hunk 6]
```diff
@@ -0,0 +1,699 @@
...
+          "nextLinkName": "nextLink"
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Network/virtualRouters": {
+      "get": {
+        "tags": [
+          "VirtualRouters"
+        ],
+        "operationId": "VirtualRouters_ListAll",
```
Comment (sanjaiganesh) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json]: >VirtualRouters_ListAll [](start = 24, length = 22)

Add resource group level collection call.. VirtualRouters_ListByResourceGroup

File: specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json [hunk 7]
```diff
@@ -0,0 +1,699 @@
...
+          }
+        },
+        "x-ms-long-running-operation": true,
+        "x-ms-long-running-operation-options": {
+          "final-state-via": "azure-async-operation"
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Network/virtualRouters/{virtualRouterName}/peerings": {
+      "get": {
```
Comment (sanjaiganesh) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json]: >get [](start = 7, length = 3)

Add both subscription and RG level collection calls

File: specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json [hunk 8]
```diff
@@ -0,0 +1,699 @@
...
+            }
+          }
+        },
+        "x-ms-examples": {
+          "Get Virtual Router Peering": {
+            "$ref": "./examples/VirtualRouterPeeringGet.json"
+          }
+        }
+      },
+      "put": {
```
Comment (sanjaiganesh) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json]: >put [](start = 7, length = 3)

add Patch api.. (min tags patch support required)

File: specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json [hunk 9]
```diff
@@ -0,0 +1,699 @@
...
+        }
+      ],
+      "description": "VirtualRouter Resource."
+    },
+    "VirtualRouterPropertiesFormat": {
+      "properties": {
+        "virtualRouterAsn": {
+          "type": "integer",
+		  "readOnly": false,
+          "format": "int64",
```
Comment (sanjaiganesh) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json]: tabs -> spaces   (before "readonly" property of virtualRouterAsn)

Comment (sanjaiganesh) [reply specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json]: same after "maximum" property

---
In reply to: [317344838](https://github.com/Azure/azure-rest-api-specs/pull/7005#discussion_r317344838) [](ancestors = 317344838)

File: specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json [hunk 10]
```diff
@@ -0,0 +1,699 @@
...
+            "required": true,
+            "type": "string",
+            "description": "The name of the resource group."
+          },
+          {
+            "name": "virtualRouterName",
+            "in": "path",
+            "required": true,
+            "type": "string",
+            "description": "The name of the Virtual Router."
```
Comment (sanjaiganesh) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json]: consider adding parameter definition for these too.. (similar to ApiVersionParameter)

File: specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json [hunk 11]
```diff
@@ -0,0 +1,699 @@
...
+      "description": "Virtual Router Peering resource"
+    },
+    "VirtualRouterPeeringProperties": {
+      "properties": {
+        "peerAsn": {
+          "type": "integer",
+		  "readOnly": false,
+          "format": "int64",
+          "minimum": 1,
+          "maximum": 4294967295,
```
Comment (sanjaiganesh) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json]: tabs -> spaces

File: specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json [hunk 12]
```diff
@@ -0,0 +1,699 @@
...
+      "description": "Properties of the rule group."
+    },
+    "VirtualRouterListResult": {
+      "properties": {
+        "value": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/VirtualRouter"
+          },
+          "description": "List of Virtual Routers in a resource group."
```
Comment (sanjaiganesh) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json]: >in a resource grou [](start = 50, length = 18)

nit - update description to not have resource group. as the List* call need not be resourcegroup level query..

File: specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json [hunk 13]
```diff
@@ -0,0 +1,699 @@
...
+            "$ref": "#/definitions/VirtualRouterPeering"
+          },
+          "description": "List of VirtualRouterPeerings in a VirtualRouter."
+        },
+        "nextLink": {
+          "type": "string",
+          "description": "URL to get the next set of results."
+        }
+      },
+      "description": "Response for VirtualRouterPeerings API service call."
```
Comment (sanjaiganesh) [inline specification/network/resource-manager/Microsoft.Network/stable/2019-07-01/virtualRouter.json]: >description [](start = 7, length = 11)

nit - update description to reflect this is a collection call for Peerings  ?

**Review by anton-evseev — CHANGES_REQUESTED** [review]

**Review by anton-evseev — CHANGES_REQUESTED** [review]

**Review by anton-evseev — CHANGES_REQUESTED** [review]

**Review by anthony-c-martin — CHANGES_REQUESTED** [review]

**Review by anthony-c-martin — APPROVED** [review]
LGTM!

**Comment by naveenchekuri** [comment]
@lirenhe and @number213 ARM team has signed off on the PR. Could you sign off as well

**Comment by number213** [comment]
@anthony-c-martin I'm still concerned about PATCH operations' `operationId`. I think if should be `_UpdateTags` since it only updates tags and that's how such operations are named in other Network specs. But previous ARM reviewer insisted on using `_Update`.

**Comment by number213** [comment]
@lirenhe could you please take a look at CI failures? Maybe `network-august-release` is missing some updates from `master`? Could you please merge master into our release branch?

**Comment by anthony-c-martin** [comment]
@number213 @naveenchekuri `_UpdateTags` sounds fine to me if it's consistent with your other API declarations, and you don't support patching any other resource properties.

**Comment by number213** [comment]
Sounds good! @naveenchekuri could you please update two operationIds to use `_UpdateTags`?

**Comment by naveenchekuri** [comment]
@number213 Done. Made changes to set as UpdateTags

**Review by anton-evseev — APPROVED** [review]

**Review by lirenhe — APPROVED** [review]

**Comment by lirenhe** [comment]
@naveenchekuri  I saw there are a couple of check errors which is a regression issue on our side. So to solve those check errors, could you sync the code from master to your working branch?

**Comment by NullMDR** [comment]
@naveenchekuri Please pick this PR https://github.com/Azure/azure-rest-api-specs/pull/7068 or merge from master to fix the CI

**Comment by number213** [comment]
Hi, @lirenhe, I've cherry picked two commit that solved CI errors. SDK java is failing due to timeout as usual, can we get this PR merged?
Also, regarding SDK java... Could we try bumping timeout in `SwaggerToSDK.yml` to 3 hours? Azure DevOps supports timeouts up to 6 hours for public pipelines, maybe that would resolve our issues.

**Comment by NullMDR** [comment]
@number123 I'm the owner of the swagger to sdk automation. I'm not aware of that the automation need to run that long of time. I need to confirm that does it really need over than an hour to generate SDK？

**Comment by number213** [comment]
Yes, as you can see, Java SDK takes more than two hours to generate because it is multi-API SDK and it tries to generate code for each version.

**Comment by number213** [comment]
@lirenhe @NullMDR gentle ping

**Comment by naveenchekuri** [comment]
@lirenhe @NullMDR. We need to commit the PR before 09. please help with this

**Comment by NullMDR** [comment]
@yungezz please force push this one. The timeout issue will be fixed in the new automation.

**Comment by number213** [comment]
Rebased PR to include fixes for CI that were merged into `network-august-branch` recently

**Comment by naveenchekuri** [comment]
@yungezz could you please merge the PR

**Pull request merged**